tests/integration_test/grpc_deadline_test.go

```diff
@@ -190,7 +192,121 @@ plans {
 
 		// Place in closure to allow efficient measuring of elapsed time.
 		// Elapsed time is not checked in the test, it's just for debugging.
-		func() {
+		t.Run(tc.desc, func(t *testing.T) {
+			defer utils.Elapsed(fmt.Sprintf("Test (%s):", tc.desc))()
+
+			// For this client, `err` will always be "exit status 1" on failures.
+			// Check for actual error in `resp` instead.
+			resp, err := client.RunGRPCEchoTest(tc.testPlan, s.Ports().ListenerPort)
+
+			if tc.wantErr == "" && err != nil {
+				t.Errorf("Test (%v): Error during running test: want no err, got err (%v)", tc.desc, resp)
+			}
+
+			if tc.wantErr != "" && err == nil {
+				t.Errorf("Test (%v): Error during running test: got no err, want err (%v)", tc.desc, tc.wantErr)
+			}
+
+			if err != nil && !strings.Contains(resp, tc.wantErr) {
+				t.Errorf("Test (%s): failed, got err (%v), expected err (%v)", tc.desc, resp, tc.wantErr)
+			}
+		})
+	}
+}
+
+func TestIdleTimeoutsForGrpcStreaming(t *testing.T) {
+	t.Parallel()
+
+	testData := []struct {
+		desc        string
+		confArgs    []string
+		addDeadline time.Duration
+		wantErr     string
+		testPlan    string
+	}{
+		// Please be cautious about adding too many time-based tests here.
+		// This can slow down our CI system if we sleep for too long.
+		{
+			desc: "When deadline is NOT specified for method, stream idle timeout specified via flag kicks in and the request fails.",
+			confArgs: append([]string{
+				"--stream_idle_timeout=15s",
+			}, utils.CommonArgs()...),
+			wantErr: `stream timeout`,
+			testPlan: `
+plans {
+  echo_stream {
+    call_config {
+      api_key: "this-is-an-api-key"
+    }
+    request {
+      text: "Hello, world!"
+      response_delay: 20
+    }
+    count: 1
+  }
+}`,
+		},
+		{
+			desc:        "When deadline is specified for method, it overrides the global stream idle timeout and the request succeeds.",
+			addDeadline: 25 * time.Second,
+			confArgs: append([]string{
+				"--stream_idle_timeout=15s",
+			}, utils.CommonArgs()...),
+			testPlan: `
+plans {
+  echo_stream {
+    call_config {
+      api_key: "this-is-an-api-key"
+    }
+    request {
+      text: "Hello, world!"
+      response_delay: 20
+    }
+    count: 1
+  }
+}`,
+		},
+		{
+			desc: "When deadline is NOT specified for method, a low stream idle timeout specified via flag is not honored and the request succeeds.",
+			confArgs: append([]string{
+				"--stream_idle_timeout=3s",
```

Comment:
This is confusing.  Naturally, people will expect this flag to apply.  

How about this:
can we distinguish between flag specified, and not-specified.  if not specified,  we want to be smart, to allow long timeout request.  
if specified,  we should honor it, always apply it

Reply:
Ok, I added a few more tests and changed the behavior a little. A few points below:

- We are not exposing the flag to the user. It is just for integration tests so we can set the default stream idle timeout. Otherwise Envoy will default to `5m`, that will be hard to integration test.
- This means the only way for the user to modify the deadlines is via service config. This will modify both the response deadline and the stream idle timeout. So let's not worry too much about the flag, only about the `deadline` from service config.

The new behavior is split into 2 cases now:
- For HTTP and unary gRPC: response timeout is set from `deadline`. Stream idle timeout does not matter. It will always be set to something higher than the user specified deadline.
- For non-unary gRPC: response timeout is disabled, it does not matter. Instead, the stream idle timeout will be set equal to the `deadline`. This will allow users to set a large or small deadline and have it override the default 5m stream idle timeout.

Reply:
I see.  Your approach is to use "deadline" from service config for two different timeouts.  For streaming, it is stream_idle_timeout, for unary, it is response_timeout.   By this way, we don't need extra config for streaming_idle_timeout. so we don't expose --stream_idle_timeout flag. It is just for testing.

Hmm, using same "deadline" for two timeout?  My concern is confusing. Maybe we can document it in openapi spec. we should be fine. 

The flag is confusing,   if we can change it to --stream_idle_timeout_for_test,  and add comment, not to expose it, I will fell better.   Its current implementation is confusing to users.

Reply:
Done. Renamed the flag. Use prefix `_test_only` for consistency.

We can document it in the `x-google-backend` docs, but I feel this new behavior is intuitive and is what uses expected from the beginning.
- Before this PR, if users set `backend.rule.deadline` for a streaming method, it would basically be ignored and the stream idle timeout would default to 5m.
- With this PR, `deadline` will work as expected for both unary and streaming methods. The default stream idle timeout of 5m will be overridden correctly in both cases.